Document Maesh and SMI (#356)
containous/maesh: docs/content/configuration.md

# Configuration

> The configuration for maesh is broken into two parts: the static configuration, and the dynamic configuration.

Reviewer: [Break into](https://idioms.thefreedictionary.com/broken+into) => _The thief breaks into a house_
[Break down into](https://dictionary.cambridge.org/dictionary/english/break-down#dataset_cbed) => _The data breaks down into three main categories._

Reviewer: ```suggestion
The configuration for maesh is broken down into two parts: the static configuration, and the dynamic configuration.
```

Please note that this value is a string, and needs to be quoted.

Further details about the rate limiting can be found [here](https://docs.traefik.io/v2.0/middlewares/ratelimit/#configuration-options)

### With Service Mesh Interface

#### Access Control

The first step is to describe what the traffic of our server application looks like.

```yaml
---
apiVersion: specs.smi-spec.io/v1alpha1
kind: HTTPRouteGroup
metadata:
  name: server-routes
  namespace: server
matches:
  - name: api
    pathRegex: /api
    methods: ["*"]
  - name: metrics
    pathRegex: /metrics
    methods: ["GET"]
```

In this example, we define a set of HTTP routes for our `server` application.

More precisely, the `server` app is composed by two routes:

- The `api` route under the `/api` path, accepting all methods

> - The `metrics` routes under the `/metrics` path, accepting only a `GET` request

Reviewer: ```suggestion
- The `metrics` routes under the `/metrics` path, accepting only `GET` requests
```

Other types of route groups and detailed information are available [in the specification](https://github.com/deislabs/smi-spec/blob/master/traffic-specs.md).

By default, all traffic is denied so we need to grant access to clients to our application. This is done by defining a `TrafficTarget`.

```yaml
---
apiVersion: access.smi-spec.io/v1alpha1
kind: TrafficTarget
metadata:
  name: client-server-target
  namespace: server
destination:
  kind: ServiceAccount
  name: server
  namespace: server
specs:
  - kind: HTTPRouteGroup
    name: server-routes
    matches:
      - api
sources:
  - kind: ServiceAccount
    name: client
    namespace: client
```

In this example, we grant access to all pods running with the service account `client` under the namespace `client` to the HTTP route `api` specified by on the group `server-routes` on all pods running with the service account `server` under the namespace `server`.

> Any client running with the service account `client` under the `client` namespace accessing `server.server.maesh/api` is allowed to access the `/api` resource. Other will receive a 404 answer from the Maesh node.

Reviewer: ```suggestion
Any client running with the service account `client` under the `client` namespace accessing `server.server.maesh/api` is allowed to access the `/api` resource. Others will receive 404 answers from the Maesh node.
```

> More information can be found [in the specification](https://github.com/deislabs/smi-spec/blob/master/traffic-access-control.md).

Reviewer: ```suggestion
More information can be found [in the SMI specification](https://github.com/deislabs/smi-spec/blob/master/traffic-access-control.md).
```

#### Traffic Splitting

> SMI defines the `TrafficSplit` resource which allows to direct incrementally a subset of the traffic to a different services.

Reviewer: ```suggestion
SMI defines the `TrafficSplit` resource which allows to direct subsets of the traffic to different services.
```

Reviewer: WDYM by `incrementally` here?

```yaml
apiVersion: split.smi-spec.io/v1alpha1
kind: TrafficSplit
metadata:
  name: server-split
  namespace server
spec:
  service: server
  backends:
  - service: server-v1
    weight: 80
  - service: server-v2
    weight: 20
```

> In this example, we define a traffic split for our server service between two version of our server, v1 and v2.

Reviewer: ```suggestion
In this example, we define a traffic split for our server service between two versions of our server, v1 and v2.
```

`server.server.maesh` directs 80% of the traffic to the server-v1 pods, and 20% of the traffic to the server-v2 pods.

> More information can be found [in the specification](https://github.com/deislabs/smi-spec/blob/master/traffic-split.md).

Reviewer: ```suggestion
More information can be found [in the SMI specification](https://github.com/deislabs/smi-spec/blob/master/traffic-split.md).
```

#### Traffic Metrics

> At the moment, Maesh does not implement the [Traffic Metrics specification](https://github.com/deislabs/smi-spec/blob/master/traffic-metrics.md).

Reviewer 2: It may be useful to link the issue for implementing it (if there's one)
